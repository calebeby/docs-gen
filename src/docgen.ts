import { SourceFile, SyntaxKind, CallExpression, Type } from 'ts-simple-ast'

type HTTPMethods = 'get' | 'put'
const acceptedMethods = ['get', 'put']

export const isHttpMethod = (call: CallExpression) =>
  call.getTypeArguments().length === 1 &&
  call
    .getChildrenOfKind(SyntaxKind.Identifier)
    .some(id => acceptedMethods.includes(id.getText()))

/*
 * Given a source file, find all exported arrow functions that include a call to get() put() or post()
 */
export const findRoutes = (file: SourceFile) =>
  file
    .getExportedDeclarations()
    .reduce<CallExpression[]>((routes, declaration) => {
      declaration
        .getChildrenOfKind(SyntaxKind.ArrowFunction)
        .forEach(arrowFunc => {
          const call = arrowFunc
            .getDescendantsOfKind(SyntaxKind.CallExpression)
            .find(isHttpMethod)
          if (call !== undefined) {
            routes.push(call)
          }
        })
      return routes
    }, [])

interface HTTPMethod {
  url: string
  responseType: Type
  method: HTTPMethods
}

export interface Get extends HTTPMethod {
  method: 'get'
}

export interface Put extends HTTPMethod {
  requestType: Type
  method: 'put'
}

export type Route = Get | Put

export const parseRoute = (call: CallExpression): Route => {
  const method = call
    .getChildrenOfKind(SyntaxKind.Identifier)
    .find(id => acceptedMethods.includes(id.getText()))
    .print()
  const responseType = call.getTypeArguments()[0].getType()
  const url = call
    .getArguments()[0]
    .getText()
    .replace(/`/g, '')
    .replace(/'/g, '')
    .replace(/\${/g, '{')
  if (method === 'get') {
    return { responseType, method, url }
  } else if (method === 'put') {
    const requestType = call.getArguments()[1].getType()
    return { responseType, requestType, method, url }
  }
}
