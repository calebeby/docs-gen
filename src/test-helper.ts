import Project from 'ts-simple-ast'

export const parseSource = (text: string) => {
  const project = new Project()
  return project.createSourceFile('file.ts', text)
}
