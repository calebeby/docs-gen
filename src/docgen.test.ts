import { SyntaxKind } from 'ts-simple-ast'
import { parseSource } from './test-helper'

import { findRoutes, parseRoute } from './docgen'

test('findRoutes', () => {
  const src = `
    export const foo = () => request<Foo>('GET', '/blah')
    export const blah = () => request<Blah>('PUT', '/asdflkasdf')
    export const asdf = () => request<Blah>('POST', '/asdflkasdf')
    export const asdf2 = () => request<Blah>('DELETE', '/asdflkasdf')
    export const asdf3 = () => request<Blah>('POST', '/something')
    const blah = () => request<Blah>('POST', '/asdflkasdf')
    `
  expect(findRoutes(parseSource(src)).map(d => d.print()))
    .toMatchInlineSnapshot(`
Array [
  "request<Foo>('GET', '/blah')",
  "request<Blah>('PUT', '/asdflkasdf')",
  "request<Blah>('POST', '/asdflkasdf')",
  "request<Blah>('DELETE', '/asdflkasdf')",
  "request<Blah>('POST', '/something')",
]
`)
})

test('parseRoute with comment', () => {
  const src = `
    /**
     * this is a leading comment
     */
    // another leading comment
    export const asdf3 = () => request<Blah>('GET', '/something')
    `
  expect(parseRoute(findRoutes(parseSource(src))[0]).comment)
    .toMatchInlineSnapshot(`
"this is a leading comment
another leading comment"
`)
})

test('parseRoute get', () => {
  const src = `
request<{name: string}>('GET', \`/blah/\${asdf}\`)
`
  const call = parseSource(src).getDescendantsOfKind(
    SyntaxKind.CallExpression,
  )[0]
  const parsedRoute = parseRoute(call)
  expect(parsedRoute.responseNode.getType().getText()).toMatchInlineSnapshot(
    `"{ name: string; }"`,
  )
  expect(parsedRoute.method).toMatchInlineSnapshot(`"GET"`)
  expect(parsedRoute.url).toMatchInlineSnapshot(`"/blah/{asdf}"`)
})

test('parseRoute put', () => {
  const src =
    'request("PUT", `/blah/${asdf}`, (d: {age: number}) => d, {name: "hello"})' // eslint-disable-line no-template-curly-in-string
  const call = parseSource(src).getDescendantsOfKind(
    SyntaxKind.CallExpression,
  )[0]
  const parsedRoute = parseRoute(call)
  expect(parsedRoute.responseNode.getType().getText()).toMatchInlineSnapshot(
    `"{ age: number; }"`,
  )
  expect(parsedRoute.requestNode.getType().getText()).toMatchInlineSnapshot(
    `"{ name: string; }"`,
  )
  expect(parsedRoute.method).toMatchInlineSnapshot(`"PUT"`)
  expect(parsedRoute.url).toMatchInlineSnapshot(`"/blah/{asdf}"`)
})

test('parseRoute deleteRequest', () => {
  const src = 'request<boolean>("DELETE", `/blah/${asdf}`)' // eslint-disable-line no-template-curly-in-string
  const call = parseSource(src).getDescendantsOfKind(
    SyntaxKind.CallExpression,
  )[0]
  const parsedRoute = parseRoute(call)
  expect(parsedRoute.responseNode.getType().getText()).toMatchInlineSnapshot(
    `"boolean"`,
  )
  expect(parsedRoute.method).toMatchInlineSnapshot(`"DELETE"`)
  expect(parsedRoute.url).toMatchInlineSnapshot(`"/blah/{asdf}"`)
})
