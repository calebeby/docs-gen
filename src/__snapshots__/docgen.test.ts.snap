// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findRoutes 1`] = `
Array [
  "get<Foo>('/blah')",
  "get<Blah>('/asdflkasdf')",
  "put<Blah>('/asdflkasdf')",
  "putRequest<Blah>('/something')",
]
`;

exports[`parseRoute deleteRequest 1`] = `"boolean"`;

exports[`parseRoute deleteRequest 2`] = `"delete"`;

exports[`parseRoute deleteRequest 3`] = `"/blah/{asdf}"`;

exports[`parseRoute get 1`] = `"any"`;

exports[`parseRoute get 2`] = `"get"`;

exports[`parseRoute get 3`] = `"/blah/{asdf}"`;

exports[`parseRoute put 1`] = `"any"`;

exports[`parseRoute put 2`] = `"{ name: string; }"`;

exports[`parseRoute put 3`] = `"put"`;

exports[`parseRoute put 4`] = `"/blah/{asdf}"`;
